protocol/decoder/decoder.go

```diff
@@ -292,14 +310,22 @@ func (d *Decoder) consumeRawBytes(length uint64) DecoderError {
 	return nil
 }
 
-func (d *Decoder) WrapError(err error) DecoderError {
-	// If we've already wrapped the error, preserve the nested path
-	if decoderError, ok := err.(*decoderErrorImpl); ok {
+func (d *Decoder) GetDecodingError(err error) *decodingError {
+	// If we've already wrapped the error, preserve the offset
+	if decoderError, ok := err.(*decodingError); ok {
 		return decoderError
 	}
 
-	return &decoderErrorImpl{
+	return &decodingError{
 		message: err.Error(),
 		offset:  int(d.buffer.Offset()),
 	}
 }
+
+func (d *Decoder) GetInvalidDecodedValueError(err error, startOffset, endOffset int) *invalidDecodedValueError {
```

Comment:
Don't add any value since we aren't using anything except the passed in arguments - instead add a constructor (`NewDecoderErrorForRange()`) and use that